src/common/connection.js

```diff
@@ -95,14 +97,44 @@ class Connection extends EventEmitter {
     return this._state === WebSocket.OPEN && this._isReady;
   }
 
-  _onUnexpectedClose(resolve = function() {}, reject = function() {}) {
+  _onUnexpectedClose(resolve, reject) {
     if (this._onOpenErrorBound) {
       this._ws.removeListener('error', this._onOpenErrorBound);
       this._onOpenErrorBound = null;
     }
     this._ws = null;
     this._isReady = false;
-    this.connect().then(resolve, reject);
+    if (_.isFunction(resolve)) {
+      // connection was closed before it was properly opened, so we must return
+      // error to connect's caller
+      this.connect().then(resolve, reject);
+    } else {
+      this.emit('disconnected', true);
+      this._retryConnect();
+    }
+  }
+
+  _retryConnect() {
+    this._retry += 1;
+    const retryTimeout = (this._retry < 40)
+      // First, for 2 seconds: 20 times per second
+      ? (1000 / 20)
```

Comment:
Can you move this logic into a separate function? The nested ternary statements are confusing to follow